From f1cbca7f3f6136f6fd69a760a815b2167d5357d9 Mon Sep 17 00:00:00 2001
From: Marc Mutz <marc.mutz@qt.io>
Date: Thu, 16 Oct 2025 11:14:10 +0200
Subject: [PATCH 438/553] =?UTF-8?q?QChar:=20avoid=20char{32=E2=86=9216}=5F?=
 =?UTF-8?q?t=20narrowing=20in=20foldCase(char16=5Ft*,=20char16=5Ft*)?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

... which triggers Clang 21 -Wcharacter-conversion.

Instead of using the char32_t ucs4 variable to temporarily store the
read from *ch, keep using *ch until we have determined whether it
needs to be decoded as a surrogate pair or can simply be promoted to
char32_t.

Since this touches most lines of the function, anyway, take the
opportunity to rename the first argument from 'ch' to 'cur', and change
pointer arithmetic to indexing operations.

Amends the start of the public history.

Pick-to: 6.8 6.5
Change-Id: I47a026fd509da8fa3dee4b6be30ac4432148d9c6
Reviewed-by: Ahmad Samir <a.samirh78@gmail.com>
(cherry picked from commit 59259fdc7362035889c075b5071c355ae41672f1)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/text/qchar.cpp | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/src/corelib/text/qchar.cpp b/src/corelib/text/qchar.cpp
index 1e4111646a8..f2bc22c421a 100644
--- a/src/corelib/text/qchar.cpp
+++ b/src/corelib/text/qchar.cpp
@@ -1678,11 +1678,13 @@ char32_t QChar::toTitleCase(char32_t ucs4) noexcept
     return convertCase_helper(ucs4, QUnicodeTables::TitleCase);
 }
 
-static inline char32_t foldCase(const char16_t *ch, const char16_t *start)
+static inline char32_t foldCase(const char16_t *cur, const char16_t *start)
 {
-    char32_t ucs4 = *ch;
-    if (QChar::isLowSurrogate(ucs4) && ch > start && QChar::isHighSurrogate(*(ch - 1)))
-        ucs4 = QChar::surrogateToUcs4(*(ch - 1), ucs4);
+    char32_t ucs4;
+    if (QChar::isLowSurrogate(*cur) && cur > start && QChar::isHighSurrogate(cur[-1]))
+        ucs4 = QChar::surrogateToUcs4(cur[-1], *cur);
+    else
+        ucs4 = *cur;
     return convertCase_helper(ucs4, QUnicodeTables::CaseFold);
 }
 
-- 
2.51.2

